train_concept_identification.py

```diff
@@ -41,6 +42,94 @@ def compute_loss(criterion, logits, gold_outputs):
+
+  Returns:
+    f_score
+  """
+
+    predicted_outputs_no_padding, gold_outputs_no_padding = tensor_to_list(gold_outputs, predicted_outputs)
+
+    ids_to_concepts_list_gold = list(vocabs.concept_vocab.keys())
+    concepts_as_list_gold = []
+    #TODO store the gold data before numericalization and use it here
+    for sentence in gold_outputs_no_padding:
+        for id in sentence:
+            if ids_to_concepts_list_gold[id] != UNK:
+                concepts_as_list_gold.append(ids_to_concepts_list_gold[id])
+
+
+    ids_to_concepts_list_predicted = list(vocabs.concept_vocab.keys())
+    concepts_as_list_predicted = []
+    for sentence in predicted_outputs_no_padding:
+        for id in sentence:
+            if ids_to_concepts_list_predicted[int(id)] != UNK:
+                concepts_as_list_predicted.append(ids_to_concepts_list_predicted[int(id)])
+
+    print("concepts_as_list_gold", concepts_as_list_gold)
+    print("concepts_as_list_predicted", concepts_as_list_predicted)
+
+    true_positive = len(set(concepts_as_list_gold) & set(concepts_as_list_predicted))
+    false_positive = len(set(concepts_as_list_predicted).difference(set(concepts_as_list_gold)))
+    false_negative = len(set(concepts_as_list_gold).difference(set(concepts_as_list_predicted)))
+    precision = 0
+    recall = 0
+    print(f'tp = {true_positive}, fn = {false_negative}, fp = {false_positive}')
+
+    if len(concepts_as_list_predicted) != 0:
+        precision = true_positive / (true_positive + false_positive)
+        recall = true_positive / (true_positive + false_negative)
+    f_score = 0
+
+    if precision + recall != 0:
+        f_score = 2 * (precision * recall) / (precision + recall)
+
+    return f_score
+
+
+def tensor_to_list(gold_outputs, predicted_outputs):
+
+    # Transpose the tensors, transform them in lists and remove the root
+    gold_list_with_padding = []
+    for sentence in torch.transpose(gold_outputs, 0, 1):
+        gold_list_with_padding.append(sentence.tolist()[1:])
+
+    predicted_list_with_padding = []
+    for sentence in torch.transpose(predicted_outputs, 0, 1):
+        predicted_list_with_padding.append(sentence.tolist()[1:])
+
+    # Remove the padding -> stop at EOS, for both gold and predicted concepts
+    gold_list_no_padding = []
+    for sentence in gold_list_with_padding:
+        sentence_no_padding = []
+        for word in sentence:
+            if word == 1:
+                break
+            else:
+                sentence_no_padding.append(word)
+        gold_list_no_padding.append(sentence_no_padding)
+
+    predicted_list_no_padding = []
+    for sentence in predicted_list_with_padding:
+        sentence_no_padding = []
+        for word in sentence:
+            if int(word) == 1:
+                break
+            else:
+                sentence_no_padding.append(word)
+        predicted_list_no_padding.append(sentence_no_padding)
```

Comment:
This code repeats ~the same thing for the predicted and gold sequence. Please extract the common functionality in a function and call it for each sequence